crypto/init.c

```diff
@@ -552,21 +485,42 @@ void OPENSSL_cleanup(void)
      * - ENGINEs and additional EVP algorithms might use added OIDs names so
      * obj_cleanup_int() must be called last
      */
+    OSSL_TRACE(INIT, "OPENSSL_cleanup: rand_cleanup_int()\n");
     rand_cleanup_int();
+
+    OSSL_TRACE(INIT, "OPENSSL_cleanup: rand_drbg_cleanup_int()\n");
     rand_drbg_cleanup_int();
+
+    OSSL_TRACE(INIT, "OPENSSL_cleanup: conf_modules_free_int()\n");
     conf_modules_free_int();
 #ifndef OPENSSL_NO_ENGINE
+    OSSL_TRACE(INIT, "OPENSSL_cleanup: engine_cleanup_int()\n");
     engine_cleanup_int();
 #endif
+    OSSL_TRACE(INIT, "OPENSSL_cleanup: ossl_store_cleanup_int()\n");
     ossl_store_cleanup_int();
+
+    OSSL_TRACE(INIT, "OPENSSL_cleanup: crypto_cleanup_all_ex_data_int()\n");
     crypto_cleanup_all_ex_data_int();
+
+    OSSL_TRACE(INIT, "OPENSSL_cleanup: bio_cleanup()\n");
     bio_cleanup();
+
+    OSSL_TRACE(INIT, "OPENSSL_cleanup: evp_cleanup_int()\n");
```

Comment:
So, uhmm, using this after the `bio_cleanup` call, how do you see that working?  I'm asking conceptually rather than with regards to what `bio_cleanup` does today.

Reply:
My understanding was that you needed the special treatment only because you were using the global `bio_err` object, whereas my proposed solution creates a dedicated channel for every category using `dup_bio_err()`. As far as I could tell, the bio methods did not rely on global objects (e.g locks) which could be invalidated by the cleanup. Also, I tested it on linux and it didn't crash ;-).

But maybe there is some detail which I was not aware of?

Reply:
You do not know what `bio_cleanup` will do in the future.  Due to its very global nature, I'm worried that, even if it's harmless to use BIOs after the call *today*, *future* changes in `bio_cleanup` may make such practice fragile, with possible mysterious crashes.

Hence that I'm asking the question on a conceptual level rather than today's specifics.